core/pow/engine.go

```diff
@@ -561,3 +576,65 @@ func (e *Engine) BlockData() (uint64, string) {
 	}
 	return e.currentBlock, e.blockHash[e.currentBlock%ringSize]
 }
+
+func (e *Engine) GetSpamStatistics(partyID string) *protoapi.PoWStatistic {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+
+	stats := make([]*protoapi.PoWBlockState, 0, len(e.activeStates)) //
+
+	for _, state := range e.activeStates {
+		for block, blockToPartyState := range state.blockToPartyState {
+			stateInd := 0
+			for i, p := range e.activeParams {
+				if block >= p.fromBlock && (p.untilBlock == nil || *p.untilBlock >= block) {
+					stateInd = i
+					break
+				}
+			}
+			params := e.activeParams[stateInd]
+			blockIndex := block % ringSize
+
+			if partyState, ok := blockToPartyState[partyID]; ok {
+				stats = append(stats, &protoapi.PoWBlockState{
+					BlockHeight:      block,
+					BlockHash:        e.blockHash[blockIndex],
+					TransactionsSeen: uint64(partyState.seenCount),
+					ExpectedDifficulty: calculateSpamExpectedDifficulty(params.spamPoWDifficulty,
+						uint(params.spamPoWNumberOfTxPerBlock),
+						partyState.seenCount,
+						partyState.observedDifficulty,
+					),
+					HashFunction: params.spamPoWHashFunction,
+				})
+			}
+		}
+	}
+
+	until := e.bannedParties[partyID].UnixNano()
+
+	var bannedUntil *string
+
+	if until > 0 {
+		untilStr := strconv.FormatInt(until, 10)
+		bannedUntil = &untilStr
+	}
+
+	return &protoapi.PoWStatistic{
+		BlockStates: stats,
+		BannedUntil: bannedUntil,
+	}
+}
+
+func calculateSpamExpectedDifficulty(spamPoWDifficulty, spamPoWNumberOfTxPerBlock, seenTx uint, observedDifficulty uint) uint64 {
```

Comment:
can you call this func getMinDifficultyForNextTx maybe? 

func getMinDifficultyForNextTx(baseDifficulty, txPerBlock, seenTx, observedDifficulty uint) uint64

Reply:
note, that none of this is correct if the boolean increaseDifficulty is false! in which case there is no difficulty that can be used for the next transaction for the block. not sure how you want to express that.